### PR TITLE
resolve where clauses in types in impls

### LINKED ISSUE #47206: where clause in generic associated types causes no output
I'm working on implementing generic associated types in the compiler with mentoring by @nikomatsakis. He recently discovered an interesting issue that we're haven't found the source of yet. It seems that having a where clause like the one below results in a compilation error but no output.

**The feature isn't actually fully finished yet, so we expect a compilation error. It just isn't clear why the compiler isn't actually reporting any specific problems. The message says that it failed due to a previous error but doesn't list any errors at all.**

Here's the code that reproduces the issue: ([Playground link](https://play.rust-lang.org/?gist=0ecb7409a3690e31d73363800c905324&version=nightly))

```rust
#![feature(generic_associated_types)]

trait Foo {
    type Assoc3<T>;
}

struct Bar;

impl Foo for Bar {
    type Assoc3<T> where T: Iterator = Vec<T>;
}

fn main() {}
```

Here's the output:

```
   Compiling playground v0.0.1 (file:///playground)
error: cannot continue compilation due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

This is on the nightly compiler because the feature is gated while we're still working on it. If you try to run this on stable or beta, you get a normal error message.

If you delete the `where T: Iterator` you get an ICE which I think is expected given what I still have left to implement for the generic associated types feature.

It might be that this gets fixed while we keep implementing generic associated types. Creating this issue will make sure that this doesn't get forgotten along the way.

### PR DESCRIPTION
**Pull request by nikomatsakis**
Fixes #47206

cc @sunjay

### CODE REVIEW

File: src/test/ui/rfc1598-generic-associated-types/issue-47206.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
```
Comment (estebank) [inline src/test/ui/rfc1598-generic-associated-types/issue-47206.rs]: 2018

**Review by sunjay — APPROVED** [review]

**Comment by sunjay** [comment]
LGTM. Thanks for looking into this! :smile:

**Review by estebank — APPROVED** [review]
r=me, feel free to disregard the nitpick

File: src/test/ui/rfc1598-generic-associated-types/issue-47206.rs (hunk 1 above)
Comment (jonas-schievink) [reply src/test/ui/rfc1598-generic-associated-types/issue-47206.rs]: Now that https://github.com/rust-lang/rust/pull/53617 was merged I think the copyright notice can just be removed.

Comment (nikomatsakis) [reply src/test/ui/rfc1598-generic-associated-types/issue-47206.rs]: I confess that I hate those licenses so much I make a point of leaving the Copyright year untouched (it doesn't matter a whit legally, from what I understand).

Comment (nikomatsakis) [reply src/test/ui/rfc1598-generic-associated-types/issue-47206.rs]: Woah, I knew nothing about #53617!

**Comment by estebank** [comment]
This change causes `unexpected panic`s in a bunch of tests :-/

I figured that it wouldn't affect any tests, let alone this many.

```

[00:47:57]     [ui] ui/codemap_tests/two_files.rs
[00:47:57]     [ui] ui/error-codes/E0404.rs
[00:47:57]     [ui] ui/error-codes/E0405.rs
[00:47:57]     [ui] ui/extern/extern-with-type-bounds.rs
[00:47:57]     [ui] ui/impl-trait/universal_wrong_bounds.rs
[00:47:57]     [ui] ui/issues/issue-2330.rs
[00:47:57]     [ui] ui/macros/macro_path_as_generic_bound.rs
[00:47:57]     [ui] ui/no-implicit-prelude-nested.rs
[00:47:57]     [ui] ui/no-implicit-prelude.rs
[00:47:57]     [ui] ui/resolve/issue-21221-1.rs
[00:47:57]     [ui] ui/resolve/issue-21221-2.rs
[00:47:57]     [ui] ui/resolve/issue-21221-3.rs
[00:47:57]     [ui] ui/resolve/issue-21221-4.rs
[00:47:57]     [ui] ui/resolve/issue-3907.rs
[00:47:57]     [ui] ui/resolve/issue-5035.rs
[00:47:57]     [ui] ui/resolve/resolve-self-in-impl-2.rs
[00:47:57]     [ui] ui/resolve/resolve-unknown-trait.rs
[00:47:57]     [ui] ui/resolve/unboxed-closure-sugar-nonexistent-trait.rs
[00:47:57]     [ui] ui/rfc1598-generic-associated-types/generic-associated-types-where.rs
[00:47:57]     [ui] ui/rfc1598-generic-associated-types/issue-47206.rs
[00:47:57]     [ui] ui/span/issue-35987.rs
[00:47:57]     [ui] ui/traits/trait-bounds-not-on-struct.rs
[00:47:57]     [ui] ui/traits/trait-test.rs
```

**Comment by kennytm** [comment]
@bors r- 

ICEs in many UI tests.

**Comment by nikomatsakis** [comment]
Fascinating. I also did not expect that. =)

**Comment by TimNN** [comment]
Ping from triage, @nikomatsakis: What are your plans for this PR?

**Comment by nikomatsakis** [comment]
Haven't had a chance to get back to it. Will soon.

**Comment by TimNN** [comment]
Ping from triage, @nikomatsakis!

**Comment by TimNN** [comment]
Ping from triage @nikomatsakis! I'm closing this PR as inactive for the moment, per https://forge.rust-lang.org/triage-procedure.html. Please feel free to re-open in the future, when you work on this again.

**Pull request closed without merging**